Firestore/core/src/firebase/firestore/immutable/append_only_list.h

```diff
@@ -0,0 +1,228 @@
+    }
+
+    // Create a the underlying vector with capacity reserved, but return the
+    // result with the current size. Reserving does not actually append anything
+    // to the underlying vector so size() shouldn't change.
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(capacity);
+    return AppendOnlyList(std::move(new_contents), size_);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the given value appended to the end.
+   *
+   * Each `push_back` creates a new instance and appears not to modify any that
+   * come. If `push_back` is called on the last instance in a chain, it will
+   * share the backing vector with the prior instance.
+   *
+   * If `push_back` is called when this instance isn't the last instance in the
+   * chain, it will make a copy of all preceding elements in the chain and
+   * return a new chain suitable for further chained `push_back` operations.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList push_back(const T& value) const {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->push_back(value);
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the given value appended to the end.
+   *
+   * @see `push_back(const T&)` for detailed discussion.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList push_back(T&& value) const {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->push_back(std::move(value));
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList constructing a new value appended to the end.
+   *
+   * @see `push_back(const T&)` for detailed discussion.
+   */
+  template <typename... Args>
+  ABSL_MUST_USE_RESULT AppendOnlyList emplace_back(Args&&... args) {
+    size_t new_size = size_ + 1;
+    std::shared_ptr<std::vector<T>> new_contents = PrepareForAppend(new_size);
+
+    new_contents->emplace_back(std::forward<Args>(args)...);
+    return AppendOnlyList(std::move(new_contents), new_size);
+  }
+
+  /**
+   * Creates a new AppendOnlyList with the final link in the chain removed.
+   *
+   * Note that the element isn't actually removed from the backing vector and
+   * it still constitutes the end of the chain. This means that any `push_back`
+   * on the resulting AppendOnlyList will result in a full copy.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList pop_back() const {
+    if (size_ <= 1) {
+      return clear();
+    }
+
+    return AppendOnlyList(contents_, size_ - 1);
+  }
+
+  /**
+   * Creates a new AppendOnlyList without any elements.
+   */
+  ABSL_MUST_USE_RESULT AppendOnlyList clear() const {
+    return AppendOnlyList(nullptr, 0);
+  }
+
+  size_t size() const {
+    return size_;
+  }
+
+  bool empty() const {
+    return size_ == 0;
+  }
+
+  const_iterator begin() const {
```

Comment:
Instead of calling this out in comments, I've changed `PrepareForAppend` to make a new copy whenever it would have reallocated the buffer backing the vector. This avoids invalidation. I've added a test (that failed initially).

Reply:
Done.